### PR TITLE
[CIS-994] Fix the issue with receiving .didBecomeActive before connecting the client for the first time

### PR DESCRIPTION
**Pull request by dmigach**
We spotted one additional case when it's possible for `ChatClientUpdater.connect` to be called before obtaining credentials:
https://github.com/GetStream/stream-chat-swift/issues/1218

It happens because we subscribe for notifications in `ChatClient`'s init
Two fixes to prevent this:
1) Do not call `ChatClientUpdater.connect` if one of  `connect` methods has not been called on `ChatClient`
2) Only listen to app state notifications in connected state, i.e. only subscribe when we connect and unsubscribe on disconnect

### CODE REVIEW

File: Sources/StreamChat/ChatClient.swift [hunk 1]
```diff
@@ -363,6 +358,11 @@ public class _ChatClient<ExtraData: ExtraDataTypes> {
     /// are received.
     public func disconnect() {
         clientUpdater.disconnect()
+        
+        backgroundTaskScheduler?.startListeningForAppStateUpdates(
```
Comment (dmigach) [inline Sources/StreamChat/ChatClient.swift:362]: Not sure if it's ok or we better add a `BackgroundTaskScheduler.stopListeningForAppStateUpdates` and incapsulate it in there

Comment (b-onc) [reply Sources/StreamChat/ChatClient.swift:362]: `stopListeningForAppStateUpdates` would be a good addition

File: Sources/StreamChat/ChatClient.swift [hunk 2]
```diff
@@ -441,6 +446,11 @@ public class _ChatClient<ExtraData: ExtraDataTypes> {
     private func handleAppDidBecomeActive() {
         cancelBackgroundTaskIfNeeded()
 
+        guard userConnectionProvider != nil else {
+            // The client has not been connected yet during this session
+            return
+        }
```
Comment (b-onc) [inline Sources/StreamChat/ChatClient.swift:452]: Isn't this enough to fix the issue? then we don't need `stopListeningForAppUpdates` or changing the place of `startListening...`

Comment (dmigach) [reply Sources/StreamChat/ChatClient.swift:452]: It is enough, but we shouldn't connect on going back active if the client was disconnected manually

Comment (b-onc) [reply Sources/StreamChat/ChatClient.swift:452]: There's no `connect` without setting user info anymore, so on explicit disconnect we can clear `userConnectionProvider`

Comment (dmigach) [reply Sources/StreamChat/ChatClient.swift:452]: This is true, though don't you think it adds an implicit assumption? Subscribe/unsubscribe is not 100% robust neither, but it's more explicit

Comment (b-onc) [reply Sources/StreamChat/ChatClient.swift:452]: the implicit assumption is made in the `guard` statement above anyway ¯\_(ツ)_/¯ up to you, both approaches seem fine at first sight

Comment (dmigach) [reply Sources/StreamChat/ChatClient.swift:452]: Yeah this is also true 🤔 So are we good with leaving both safety mechanisms in place? I'm adding `stopListeningForAppStateUpdates`, can set `userConnectionProvider ` to `nil` on disconnecting as well

Comment (b-onc) [reply Sources/StreamChat/ChatClient.swift:452]: fine by me

File: Sources/StreamChat/ChatClient.swift [hunk 3]
```diff
@@ -413,6 +410,11 @@ public class _ChatClient<ExtraData: ExtraDataTypes> {
             userConnectionProvider: userConnectionProvider,
             completion: completion
         )
+        
+        backgroundTaskScheduler?.startListeningForAppStateUpdates(
+            onEnteringBackground: { [weak self] in self?.handleAppDidEnterBackground() },
+            onEnteringForeground: { [weak self] in self?.handleAppDidBecomeActive() }
+        )
```
Comment (VojtaStavik) [inline Sources/StreamChat/ChatClient.swift:417]: Shouldn't this happen only if the connection is successful?

File: Sources/StreamChat/ChatClient_Tests.swift [hunk 4]
```diff
@@ -858,6 +862,7 @@ class ChatClient_Tests: StressTestCase {
             eventWorkerBuilders: [],
             environment: testEnv.environment
         )
+        client.connectAnonymousUser()
```
Comment (VojtaStavik) [inline Sources/StreamChat/ChatClient_Tests.swift:865]: why this change?

**Review by VojtaStavik — APPROVED** [review]

File: Sources/StreamChat/ChatClient_Tests.swift (hunk 4 above)
Comment (dmigach) [reply Sources/StreamChat/ChatClient_Tests.swift:865]: We need to connect before we can simulate going to background. Basically all the existing tests that test the way we handle config settings and implementation details of going background/active now need to connect first

File: Sources/StreamChat/ChatClient.swift (hunk 3 above)
Comment (dmigach) [reply Sources/StreamChat/ChatClient.swift:417]: I think you're right, there is one interesting case when we go background not having connected yet, but current implementation ignores it anyway
```
    private func handleAppDidEnterBackground() {
        // We can't disconnect if we're not connected
        guard connectionStatus == .connected else { return }
    ...
```
